Reflectometry GUI restructuring 1: Settings tabs

gemmaguest opened the pull request:
**Description of work.**
Ongoing work to integrate the `JobTreeView` into the `ISIS Reflectometry` GUI. This PR is being merged to a feature branch because not all functionality is implemented yet.

This PR includes replacement of the `Settings` tab into separate `Event` and `Instrument` tabs. The new tabs use MVP models and improved validation.

This PR has a large number of additions and deletions due to a lot of code having been moved around. In particular:
- the contents of `Presenters/` has been moved to `GUI/RunsTable/`, with the `Batch` components being renamed to `JobsTable`
- the `Event` presenters/views have been replaced by the new components in `GUI/Event`
- the `SettingsTab` components have been removed and replaced by two new components, `GUI/Experiment` and `GUI/Instrument`.
- Most of the slicing code has been removed as this is no longer required (it doesn't belong in the GUI logic so has been moved to an algorithm which we will call once we implement processing - not done yet)
- The `ReflDataProcessor...` items have been removed as the data processor widget has been replaced by the JobTreeView.

**Report to:** nobody

**To test:**

Note that this is work in progress on a feature branch so not all functionality is available. Here are some suggestions for tests:
- Open the `ISIS Reflectometry` interface.
- Try the buttons above the table to add/delete rows/groups
- Try entering values in the table and copy/paste etc.
- Try keyboard shortcuts like Enter to add a row and Ctrl-Enter to add a child row
- Try entering an investigation ID e.g. `INTER` and `1120015` and click `Search`. Note that catalog login is not currently implemented so you need to first login to the catalog via the main Catalog menu.
- The search should succeed and show all the runs for the investigation. Select some of them and click `Transfer`. They should be added to the main table and grouped by title with runs for different angles on different rows. If runs have the same title and angle they will appear on the same line as summed, e.g. `11935+11936`
- Click `Autoprocess`. This will copy all runs for the experiment into the main table and start an endless progress bar. Otherwise it is not implemented and currently does nothing. Note that Pause is not currently implemented.
- Click `Start monitor`. This should start live data monitoring for the current instrument. Once it's running `Stop monitor` should be available to stop it. Note that this does *not* currently pick up settings from the settings tabs so will just run with the algorithm defaults.
- Try editing inputs on the Event, Experiment and Instrument tabs. These values don't currently do anything because processing is not implemented, but check that e.g. invalid values are disallowed or validated appropriately
- Check that the Save tab works - load a workspace and click `Refresh` to see it in the GUI. Try changing some properties and clicking `Save`. It should be saved in the given save path.
- Try clicking `Batch 2` on the vertical tabs on the left. This gives a new, blank set of horizontal tabs where you can enter entirely different set of settings. Try editing some things and toggling back to the original batch.
- Try adding new batches from the `Batch` menu and deleting batches with the close buttons on the tabs.

** Known issues **
- There is a crash if you delete the last row in the table.
- There is a known crash bug if you press Ctrl-Enter more than once quickly
- If you copy a group then paste it onto a child row of an existing group, the tree extends to more than 2 levels deep (it should be 2 max)
- Processing and Autoprocess are not implemented
- Catalog login is not implemented
- The changes from #23549 to one of the deleted files have not been incorporated in the new files yet. I'll do this as a separate Pr.

Fixes #23043 and Refs #22263

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

mantid-builder commented:
@gemmaguest the build is failing, can you investigate.
<!-- {"problem_type": "failing"} -->

mantid-builder commented:
@gemmaguest the build is failing, can you investigate.
<!-- {"problem_type": "failing"} -->

mantid-builder commented:
@gemmaguest the build is failing, can you investigate.
<!-- {"problem_type": "failing"} -->

mantid-builder commented:
@gemmaguest the build is failing, can you investigate.
<!-- {"problem_type": "failing"} -->

gemmaguest commented:
I've marked this as In Progress again while I update the release notes and merge latest changes from master

gemmaguest commented:
The system test failure is unrelated so this is ready to look at

mantid-builder commented:
@Mantid-Matthew do you have any comments on this PR?
<!-- {"problem_type": "review_requested"} -->

Mantid-Matthew commented:
Having had a brief test I have run into the following issues
 * I find that after clicking on start monitor that stop monitor is not enabled. 
 * In instrument settings I find I can enter -0 but no other negative numbers, i'm not sure how much this matters though
 * Tooltips for entry boxes are really good :+1:
 * I also ran into a few hard crashes/uncuaght exceptions which you may already be aware of. These happened when: deleting last group; trying once to delete multiple groups; when closing the GUI. I struggled to replicate any but the first however.

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/Experiment/ExperimentView.cpp:
```diff
@@ -0,0 +1,574 @@
+  /// TODO This works but also changes the font
+  spinBox.setStyleSheet("QDoubleSpinBox { background-color: #ffb8ad; }");
+
+  /// This doesn't work
```
I've checked all the TODOs and all apart from these spin box ones they will all be addressed in a future PR. The spinboxes are a minor issue so I've removed the TODO from these and just noted the problem as a comment.

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/Experiment/ExperimentView.cpp:
```diff
@@ -0,0 +1,574 @@
+
+void ExperimentView::showStitchParametersInvalid() {
+  auto palette = stitchOptionsLineEdit().palette();
+  palette.setColor(QPalette::Base, QColor("#ffb8ad"));
```
Yes, this would be better, but I'd like to do it in a future PR. I've made a note on #23051 to do this.

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/Instrument/InstrumentView.cpp:
```diff
@@ -0,0 +1,293 @@
+}
+
+template <typename T, typename StringConverter>
+std::string toCsv(std::vector<T> const &values, StringConverter toString) {
```
I think this would be better but I'd prefer to do it in a future PR. I've made a note on #23051

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/Reduction/ReductionJobs.cpp:
```diff
@@ -6,341 +12,137 @@
-      });
-  return UnslicedReductionJobs(std::move(unslicedGroups));
+void prettyPrintModel(ReductionJobs const & /*jobs*/) {
+  std::cout << "ReductionJobs:\n";
```
This was only used for debug output so I've removed it

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/Reduction/Row.h:
```diff
@@ -1,120 +1,60 @@
-using SlicedRow = Row<SlicedReductionWorkspaces>;
-using UnslicedRow = Row<ReductionWorkspaces>;
-using RowVariant = boost::variant<SlicedRow, UnslicedRow>;
+// std::ostream &operator<<(std::ostream &os, Row const &row) {
```
I think it was debug output so I've removed it

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/ReflRunsTabPresenter.cpp:
```diff
@@ -281,24 +240,24 @@ void ReflRunsTabPresenter::populateSearch(IAlgorithm_sptr searchAlg) {
  */
 void ReflRunsTabPresenter::startNewAutoreduction() {
+  if (requireNewAutoreduction()) {
+    // If starting a brand new autoreduction, delete all rows / groups in
```
Yes

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/IReflMainWindowPresenter.h:
```diff
@@ -21,56 +21,12 @@ IReflMainWindowPresenter is the interface defining the functions that the main
-  /// Set the instrument name
-  virtual void setInstrumentName(const std::string &instName) const = 0;
-  /// Data processing check for all groups
+  //  virtual void setInstrumentName(const std::string &instName) const = 0;
```
Yes, I've already done this in a subsequent PR (it's actually been moved to the batch presenter)

gemmaguest commented:
@Mantid-Matthew Thanks for the comments. I think I've addressed all the in-line comments.

Regarding the issues found during testing:
- `Stop monitor` will only be enabled if the live data monitor starts up ok. Out of cycle typically this will fail and it can take a long time for `StartLiveData` to time out (too long really, but this is also a problem on master). It should time out eventually and `Start monitor` will be re-enabled.
- Both `0` and `-0` get changed to `Unset` for me - do you also see this? I agree it's a bit weird to be able to enter `-0` but I don't think it matters if it shows as `Unset`. If it shows as `-0` though that's a bit confusing.
- The crash on deleting the last group is fixed in a subsequent PR as well as a crash pressing Ctrl-enter multiple times. I haven't seen the other crashes, but stability seems a lot better since these two fixes. I've added an item on #23051 to do some thorough stability testing before the feature branch gets merged to master.

martyngigg commented on qt/scientific_interfaces/ISISReflectometry/CMakeLists.txt:
```diff
@@ -1,100 +1,80 @@
+add_subdirectory(GUI)
+add_subdirectory(Reduction)
+
+message(${GUI_SRC_FILES})
```
As this is going to an integration branch I won't force this to go back round again but could we make a note to remove this print statement before it gets to master. The CMake output is already quite crowded.
